diff -uNr gitosis/access.py gitosis/access.py
--- gitosis/access.py	2009-09-17 08:32:59.000000000 +0300
+++ gitosis/access.py	2012-05-23 16:49:05.141290001 +0300
@@ -1,8 +1,18 @@
-import os, logging
+import os, logging, re
 from ConfigParser import NoSectionError, NoOptionError

 from gitosis import group

+def isCqdeGitProject(project_name, conf):
+    log = logging.getLogger('gitosis.access.isCqdeGitProject')
+    project_config = conf.getEnvironmentConfigPath(project_name)
+    with open(project_config, 'r') as f:
+        for line in f:
+            if re.match(r"^\s*repository_type\s*=\s*git\s*$", line):
+                return True
+    log.debug('Version control type of CQDE project %s is not git.' % project_name)
+    return False
+
 def haveAccess(config, user, mode, path):
     """
     Map request for write access to allowed path.
@@ -86,3 +96,57 @@
                 path=mapping,
                 ))
             return (prefix, mapping)
+
+    # If the access was not granted by the gitosis.conf settings,
+    # we will get to this point.
+    # Check for the authentication from CQDE system.
+    from multiproject.core.util import env_id, resolve_project_id
+    from multiproject.core.permissions import CQDEPermissionPolicy
+    from multiproject.core.configuration import conf
+    from multiproject.core.proto import ProtocolManager
+
+    if path.startswith('git/'):
+        path = path[4:]
+
+    project_name = path
+    log.debug('CQDE access check for %(user)r path %(path)r' %
+              dict(user=user, path=project_name,))
+
+    # First check if the project is a CQDE Git project.
+    if not isCqdeGitProject(project_name, conf):
+        return None
+
+    # Map the mode to the action.
+    action = None
+    if mode == 'readonly':
+        action = 'VERSION_CONTROL_VIEW'
+    elif mode == 'writable' or mode == 'writeable':
+        action = 'VERSION_CONTROL'
+    else:
+        return None
+
+    # Get the id of the project and check if the user has permissions
+    # for the action on the project.
+    env_id = env_id(project_name)
+    project_id = resolve_project_id(project_name)
+    policy = CQDEPermissionPolicy()
+    has_permission = policy.check_permission(env_id, action, user)
+
+    protocols = ProtocolManager(project_id)
+    if not protocols.is_protocol_allowed('ssh', 'git'):
+        return None
+
+    if has_permission:
+        log.debug('CQDE Access ok for %(user)r as %(mode)r on %(path)r' %
+                  dict(user=user, mode=mode, path=path,))
+
+        # Get the prefix and mapping
+        vcs_path = conf.getEnvironmentVcsPath(project_name)
+        if vcs_path:
+            mapping = path
+            prefix = os.path.dirname(vcs_path)
+            log.debug('Using prefix %(prefix)r for %(path)r' %
+                      dict(prefix=prefix, path=mapping,))
+            return (prefix, mapping)
+    else:
+        log.info('No CQDE access (%s - %s - %s)' % (user, mode, path))
diff -uNr gitosis/init.py gitosis/init.py
--- gitosis/init.py	2009-09-17 08:32:59.000000000 +0300
+++ gitosis/init.py	2011-02-25 14:41:03.943075000 +0200
@@ -129,7 +129,7 @@
         log.info('Creating repository structure...')
         repositories = util.getRepositoryDir(cfg)
         util.mkdir(repositories)
-        admin_repository = os.path.join(repositories, 'gitosis-admin.git')
+        admin_repository = os.path.join(repositories, 'gitosis-admin') # CQDE change: removed .git extension.
         init_admin_repository(
             git_dir=admin_repository,
             pubkey=pubkey,
diff -uNr gitosis/run_hook.py gitosis/run_hook.py
--- gitosis/run_hook.py	2009-09-17 08:32:59.000000000 +0300
+++ gitosis/run_hook.py	2012-05-23 17:06:32.414827211 +0300
@@ -34,14 +34,17 @@
     gitweb.set_descriptions(
         config=cfg,
         )
-    generated = util.getGeneratedFilesDir(config=cfg)
-    gitweb.generate_project_list(
-        config=cfg,
-        path=os.path.join(generated, 'projects.list'),
-        )
-    gitdaemon.set_export_ok(
-        config=cfg,
-        )
+    # Multiproject does not need generated projects list, or git-daemon
+    # to access the repositories. Accesses are checked in a custom way
+    # where the handling of the repository is left for gitosis.
+    # generated = util.getGeneratedFilesDir(config=cfg)
+    # gitweb.generate_project_list(
+    #     config=cfg,
+    #     path=os.path.join(generated, 'projects.list'),
+    #     )
+    # gitdaemon.set_export_ok(
+    #     config=cfg,
+    #     )
     authorized_keys = util.getSSHAuthorizedKeysPath(config=cfg)
     ssh.writeAuthorizedKeys(
         path=authorized_keys,
diff -uNr gitosis-orig/gitosis/serve.py gitosis-changed/gitosis/serve.py
--- gitosis/serve.py	2009-09-17 08:32:59.000000000 +0300
+++ gitosis/serve.py	2011-02-25 14:41:03.943075000 +0200
@@ -128,7 +128,9 @@
     (topdir, relpath) = newpath
     assert not relpath.endswith('.git'), \
            'git extension should have been stripped: %r' % relpath
-    repopath = '%s.git' % relpath
+    #repopath '%s.git' % relpath    # CQDE does not use .git extension for
+    repopath = relpath              # its project repositories.
+
     fullpath = os.path.join(topdir, repopath)
     if not os.path.exists(fullpath):
         # it doesn't exist on the filesystem, but the configuration
diff -uNr gitosis/ssh.py gitosis/ssh.py
--- gitosis/ssh.py	2009-09-17 08:32:59.000000000 +0300
+++ gitosis/ssh.py	2011-02-25 14:41:03.943075000 +0200
@@ -34,14 +34,14 @@
 COMMENT = '### autogenerated by gitosis, DO NOT EDIT'

 def generateAuthorizedKeys(keys):
-    TEMPLATE=('command="gitosis-serve %(user)s",no-port-forwarding,'
+    TEMPLATE=('command="GIT_COMMITTER_NAME=%(user)s gitosis-serve %(user)s",no-port-forwarding,'
               +'no-X11-forwarding,no-agent-forwarding,no-pty %(key)s')

     yield COMMENT
     for (user, key) in keys:
         yield TEMPLATE % dict(user=user, key=key)

-_COMMAND_RE = re.compile('^command="(/[^ "]+/)?gitosis-serve [^"]+",no-port-forw'
+_COMMAND_RE = re.compile('^command="[^"]+ (/[^ "]+/)?gitosis-serve [^"]+",no-port-forw'
                          +'arding,no-X11-forwarding,no-agent-forwardi'
                          +'ng,no-pty .*')

