--- browser.py.old	2013-06-17 18:34:42.575887058 +0300
+++ browser.py	2013-06-17 18:39:59.341102704 +0300
@@ -370,7 +370,7 @@
                 rev_or_latest = rev or repos.youngest_rev
                 node = get_existing_node(req, repos, path, rev_or_latest)
             except NoSuchChangeset, e:
-                raise ResourceNotFound(e.message,
+                raise ResourceNotFound("No corresponding changeset found in the repository",
                                        _('Invalid changeset number'))
 
             context = context(repos.resource.child('source', path,
